### PR TITLE
Implement Deutsch-Jozsa Algorithm In Qiskit

### PR DESCRIPTION
**Pull request by abhishekjiitr**
Signed-off-by: Abhishek Jaisingh <abhi2254015@gmail.com>

### **Describe your change:**
Implement Deutsch-Jozsa Algorithm In Qiskit 

* [x] Add an algorithm?
### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by abhishekjiitr** [comment]
Implemented the Deutsch-Jozsa Algorithm, one of the first examples of a quantum algorithm that is exponentially faster than any possible deterministic classical algorithm
@cclauss please review

File: quantum/deutsch_jozsa.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+"""
+Deutsch-Josza Algorithm is one of the first examples of a quantum
+algorithm that is exponentially faster than any possible deterministic
+classical algorithm
+
+Premise:
+We are given a hidden Boolean function  f ,
```
Comment (cclauss) [inline quantum/deutsch_jozsa.py:8]:
```suggestion
We are given a hidden Boolean function f,
```

Comment (abhishekjiitr) [reply quantum/deutsch_jozsa.py]: fixed extra whitespace

**Comment by dhruvmanila** [comment]
@cclauss What's the deal with this GitHub action which applies `black` to the code? Why doesn't it work on every other PR as well? According to you, [it does not work](https://github.com/cclauss/autoblack#tldr-it-does-not-work).

**Comment by cclauss** [comment]
The thing that it does do is...  If a maintainer (with write privileges to the repo) creates a PR then that PR blackens the entire repo.  This ensures that from time to time the entire repo is blackened even if someone lands a PR that is not green.

Go ahead and make the changes that you suggest above.

File: quantum/deutsch_jozsa.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    # plus one output qubit
+    oracle_qc = q.QuantumCircuit(n + 1)
+
+    # First, let's deal with the case in which oracle is balanced
+    if case == "balanced":
+        # First generate a random number that tells us which CNOTs to
+        # wrap in X-gates:
+        b = np.random.randint(1, 2 ** n)
+        # Next, format 'b' as a binary string of length 'n', padded with zeros:
+        b_str = format(b, "0" + str(n) + "b")
```
Comment (cclauss) [inline quantum/deutsch_jozsa.py:44]: Please use a f-string.

File: quantum/deutsch_jozsa.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        # Next, we place the first X-gates. Each digit in our binary string
+        # correspopnds to a qubit, if the digit is 0, we do nothing, if it's 1
+        # we apply an X-gate to that qubit:
+        for qubit in range(len(b_str)):
```
Comment (cclauss) [inline quantum/deutsch_jozsa.py:48]: Please use enumerate() instead of range(len()).

Comment (cclauss) [reply quantum/deutsch_jozsa.py:48]: qubit is not a qubit, it is an int.  Maybe call it I or index.

File: quantum/deutsch_jozsa.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            if b_str[qubit] == "1":
+                oracle_qc.x(qubit)
+        # Do the controlled-NOT gates for each qubit, using the output qubit
+        # as the target:
+        for qubit in range(n):
+            oracle_qc.cx(qubit, n)
+        # Next, place the final X-gates
+        for qubit in range(len(b_str)):
```
Comment (cclauss) [inline quantum/deutsch_jozsa.py:56]: enumerate()

File: quantum/deutsch_jozsa.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+References:
+- https://en.wikipedia.org/wiki/Deutsch-Jozsa_algorithm
+- https://qiskit.org/textbook/ch-algorithms/deutsch-jozsa.html
+"""
+
+import numpy as np
+import qiskit as q
+
+
+def dj_oracle(case: str, n: int) -> q.QuantumCircuit:
```
Comment (cclauss) [inline quantum/deutsch_jozsa.py:28]: Perhaps this should be two separate functions.

Comment (cclauss) [reply quantum/deutsch_jozsa.py:28]: The variable name n is not self documenting.  Can we come up with a more descriptive name.

**Review by cclauss — CHANGES_REQUESTED** [review]

File: quantum/deutsch_jozsa.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    # Return the histogram data of the results of the experiment.
+    return job.result().get_counts(dj_circuit)
+
+
+if __name__ == "__main__":
+    counts = deutsch_jozsa("constant", 3)
+    print(f"Deutsch Jozsa - Constant Oracle: {counts}")
+
+    counts = deutsch_jozsa("balanced", 3)
+    print(f"Deutsch Jozsa - Balanced Oracle: {counts}")
```
[the remark this answers is not in the document]
Comment (abhishekjiitr) [reply quantum/deutsch_jozsa.py]: fixed :+1:

File: quantum/deutsch_jozsa.py (hunk 2 above)
Comment (abhishekjiitr) [reply quantum/deutsch_jozsa.py]: switched to f-string

File: quantum/deutsch_jozsa.py (hunk 3 above)
Comment (abhishekjiitr) [reply quantum/deutsch_jozsa.py]: using _enumerate_ now, changed var name to _index_

File: quantum/deutsch_jozsa.py (hunk 4 above)
Comment (abhishekjiitr) [reply quantum/deutsch_jozsa.py]: fixed

File: quantum/deutsch_jozsa.py (hunk 5 above)
Comment (abhishekjiitr) [reply quantum/deutsch_jozsa.py]: changed var name from n to _num_qubits_

Comment (abhishekjiitr) [reply quantum/deutsch_jozsa.py]: However, I feel the function is fine as it is.
I think we shouldn't split the function.

**Review by cclauss — APPROVED** [review]
RSLGTM

**Pull request merged**